app/services/provider_interface/sort_application_choices.rb

```diff
@@ -1,24 +1,101 @@
 module ProviderInterface
   class SortApplicationChoices
-    def self.call(application_choices:, sort_by:)
-      application_choices.order(sort_order(sort_by))
+    def self.call(application_choices:)
+      for_task_view(application_choices).order(sort_order)
     end
 
-    def self.sort_order(sort_by)
-      return { updated_at: :desc } if sort_by != 'days_left_to_respond'
+    def self.for_task_view(application_choices)
+      application_choices.from <<~WITH_TASK_VIEW_GROUP.squish
+        (
+          SELECT a.*, c.recruitment_cycle_year,
+            CASE
+              WHEN #{deferred_offers_pending_reconfirmation} THEN 1
+              WHEN #{previous_cycle_pending_conditions} THEN 2
+              WHEN #{about_to_be_rejected_automatically} THEN 3
+              WHEN #{awaiting_provider_decision} THEN 4
+              WHEN #{waiting_on_candidate} THEN 5
+              ELSE 999
+            END AS task_view_group,
+            #{pg_days_left_to_respond} AS pg_days_left_to_respond
+
+            FROM application_choices a
+            LEFT JOIN courses c ON c.id = a.course_option_id
+        ) AS application_choices
+      WITH_TASK_VIEW_GROUP
+    end
+
+    def self.deferred_offers_pending_reconfirmation
+      # TODO: this query depends on deferrals implementation
+      <<~DEFERRED_OFFERS_PENDING_RECONFIRMATION.squish
+        (
+          status='deferred'
+            AND c.recruitment_cycle_year = #{RecruitmentCycle.current_year}
+        )
+      DEFERRED_OFFERS_PENDING_RECONFIRMATION
+    end
 
-      Arel.sql(
-        <<-ORDER_BY.strip_heredoc,
+    def self.previous_cycle_pending_conditions
+      <<~PREVIOUS_CYCLE_PENDING_CONDITIONS.squish
         (
-          CASE
-            WHEN (status='awaiting_provider_decision' AND (DATE(reject_by_default_at) > '#{Time.zone.now.iso8601}')) THEN 1
-            ELSE 0
-          END
-        ) DESC,
-        reject_by_default_at ASC,
-        application_choices.updated_at DESC
-        ORDER_BY
-      )
+          status='pending_conditions'
+            AND c.recruitment_cycle_year = #{RecruitmentCycle.previous_year}
+        )
+      PREVIOUS_CYCLE_PENDING_CONDITIONS
+    end
+
+    def self.about_to_be_rejected_automatically
+      <<~AWAITING_PROVIDER_DECISION.squish
+        (
+          status='awaiting_provider_decision'
+            AND c.recruitment_cycle_year = #{RecruitmentCycle.current_year}
```

Comment:
This makes sense bc everything is RBD'd before we start the new cycle, right?

<img width="1368" alt="Screenshot 2020-08-19 at 09 43 28" src="https://user-images.githubusercontent.com/642279/90612813-77711b00-e200-11ea-810c-8e5c168b12c3.png">

Reply:
I guess! I think we can scrutinise this further when implementing TaskView for real.